csrc/custom_all_reduce.cuh

```diff
@@ -0,0 +1,542 @@
+      if (i != rank_) {
+        char *handle;
+        CUDACHECK(cudaIpcOpenMemHandle(
+            (void **)&handle, *((const cudaIpcMemHandle_t *)handles[i].data()),
+            cudaIpcMemLazyEnablePeerAccess));
+        ipc_handles_.push_back(handle);
+        handle += offsets[i];
+        data.ptrs[i] = handle;
+      } else {
+        data.ptrs[i] = self;
+      }
+    }
+    auto d_data = d_rank_data_base_++;
+    CUDACHECK(
+        cudaMemcpy(d_data, &data, sizeof(RankData), cudaMemcpyHostToDevice));
+    buffers_[self] = d_data;
+  }
+
+  // note: when registering graph buffers, we intentionally choose to not
+  // deduplicate the addresses. That means if the allocator reuses some
+  // addresses, they will be registered again. This is to account for the remote
+  // possibility of different allocation patterns between ranks. For example,
+  // rank 1 may get the same input address for the second allreduce, but rank 2
+  // got a different address. IPC handles have internal reference counting
+  // mechanism so overhead should be small.
+  void register_graph_buffers(
+      const std::vector<std::string> &handles,
+      const std::vector<std::vector<int64_t>> &offsets) {
+    auto num_buffers = graph_unreg_buffers_.size();
+    check_rank_data_capacity(num_buffers);
+    std::vector<RankData> rank_data(num_buffers);
+    for (int i = 0; i < num_buffers; i++) {
+      auto self_ptr = graph_unreg_buffers_[i];
+      auto &rd = rank_data[i];
+      for (int j = 0; j < world_size_; j++) {
+        if (j != rank_) {
+          char *handle;
+          CUDACHECK(cudaIpcOpenMemHandle(
+              (void **)&handle,
+              *((cudaIpcMemHandle_t *)&handles[j]
+                                              [i * sizeof(cudaIpcMemHandle_t)]),
+              cudaIpcMemLazyEnablePeerAccess));
+          ipc_handles_.push_back(handle);
+          handle += offsets[j][i];
+          rd.ptrs[j] = handle;
+        } else {
+          rd.ptrs[j] = self_ptr;
+        }
+      }
+    }
+    CUDACHECK(cudaMemcpy(d_rank_data_base_, rank_data.data(),
+                         sizeof(RankData) * num_buffers,
+                         cudaMemcpyHostToDevice));
+    d_rank_data_base_ += num_buffers;
+    graph_unreg_buffers_.clear();
+  }
+
+  // note: 512, 36 is good for most cases
+  template <typename T>
+  void allreduce(cudaStream_t stream, T *input, T *output, int size,
+                 int threads = 512, int block_limit = 36) {
+    auto d = packed_t<T>::P::size;
+    if (size % d != 0)
+      throw std::runtime_error(
+          "custom allreduce currently requires input length to be multiple of " +
+          std::to_string(d));
+
+    RankData *ptrs;
+    cudaStreamCaptureStatus status;
+    CUDACHECK(cudaStreamIsCapturing(stream, &status));
+    if (status == cudaStreamCaptureStatusActive) {
+      ptrs = d_rank_data_base_ + graph_unreg_buffers_.size();
+      graph_unreg_buffers_.push_back(input);
+    } else {
```

Comment:
I found this part quite confusing.

The following is my understanding: in the eager mode, the custom kernel requires memcpy-ing the input tensor to a pre-registered buffer before executing the kernel. However, we want to avoid the overhead when using CUDA graph, by directly using the input tensor inside the all reduce kernel. This can be implemented by 1) using a dummy `RankData` at the capturing pass while recording the input tensor addresses, and 2) copying the recorded tensor addresses into the dummy `RankData` captured into the graph.

Is my understanding correct? If so, is it safe to use a dummy `RankData` during the graph capture?

Reply:
Also, I feel keeping track of `d_rank_data_base_` adds a quite bit of complexity into the code. Can we somehow simplify the logic?

Reply:
I was wondering if using the uninitialized `d_rank_data_base_` as the input of the all reduce kernel during capturing can be somehow unsafe because the values in `d_rank_data_base_` are not real IPC addresses. After second thought, I realized that this is not unsafe since GPU does not actually execute the all reduce kernel at the capturing time. Thanks for the clarification!